Piecewise orbit model

poneill129 opened the pull request:
This is intended as an extension to the existing BT model currently implemented. BT Piecewise was designed to allow a user to fit for orbital phase (T0) with user-defined MJD blocks. 

Preliminary work was made to allow for other piecewise orbital parameters, but since this was a tool intended for a specific problem, the development of other piecewise orbital parameters was stopped. Though the foundations for future development in this area are present.

It feels like setting the piecewise parameters from binary_piecewise to BT_piecewise is currently a jury-rig which could do with improvements. Additional concerns surround this involves when a setup() function should be called. A better understanding of this would help performance time.

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+        # self.validate()
+
+    def update_binary_object(self, toas=None, acc_delay=None):
+        super().update_binary_object(toas=toas, acc_delay=acc_delay)
```
If all this method does is call the same method on `super()` can it not be omitted completely? Several other methods here seem to be in the same situation.

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+        self.validate()
+
+    def add_piecewise_param(self, param, param_unit, paramx, j):
+        if int(j) in self.get_prefix_mapping_component("X_"):
```
This should be after the `if j is None`, shouldn't it?

poneill129 replied:
Reordered the statements, good catch

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+        if param is "A1":
+            self.add_param(
+                prefixParameter(
+                    name=param + "X_{0}".format(i),
```
Nicer to use `f"X_{i}"`

poneill129 replied:
Changed string formatting as suggested (here and in other places).

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+                    frozen=False,
+                )
+            )
+        elif param is "T0":
```
For strings it's better to use `==` than `is` - you could have two different strings that contain `T0` that were equal but not `is`.

poneill129 replied:
changed all instances of is to == for strings

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+
+    def setup(self):
+        # print(self.PB)
+        super(BinaryBTPiecewise, self).setup()
```
Just use `super().setup()`.

poneill129 replied:
Changed to `super().setup()`

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+        # if None in T0Xs.values():
+        # print("Group with non-defined T0X, applying default T0 to group")
+        # TODO set default T0 value
+        if None not in T0Xs.values():
```
Doesn't this skip defining all the parameters if any of the values is unset? That doesn't seem right?

aarchiba commented on src/pint/models/binary_piecewise.py:
```diff
@@ -0,0 +1,325 @@
+                self.binary_instance.add_binary_params(xr2_name, xr2_value)
+        self.update_binary_object()
+
+    def validate(self):
```
Are you sure there aren't any ways the new parameters could be set that would make the model unusable? `validate()` should check for invalid settings of the piecewise parameters. (Overlapping T0X intervals? T0X intervals with end before beginning? Any weirdness that won't work can be detected here and then you won't need to worry about it later.)

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        self.extended_group_range = []
+        self.d_binarydelay_d_par_funcs = [self.d_BTdelay_d_par]
+        if t is not None:
+            self._t = t
```
It's probably a good idea to set `selt._t` to None if `t` is None, rather than leaving it uninitialized.

poneill129 replied:
Changed as suggested

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if input_params is not None:
+            if self.T0X is None:
+                self.update_input(input_params)
+            elif self.T0X is not None:
```
This is just `else` isn't it?

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+            if self.T0X is None:
+                self.update_input(input_params)
+            elif self.T0X is not None:
+                self.update_input()
```
Isn't this doing exactly the same thing in both branches of the `if` statement?

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        super().set_param_values(valDict=valDict)
+        self.piecewise_parameter_loader(valDict=valDict)
+
+    def piecewise_parameter_loader(self, valDict=None):
```
It's preferable if a function name is an imperative - `load_piecewise_parameters`, perhaps, or `setup_internal_structures` or something. This function should have a docstring so you can explain what it does and when to call it.

poneill129 replied:
Forgot to add the docstring with the update, still to do...

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if (
+            valDict is None
+        ):  # If there are no updates passed by binary_instance, sets default value (usually overwritten when reading from parfile)
+            self.T0X_arr = self.T0
```
Does this make `T0X_arr` into a scalar instead of an array? Won't this break things?

poneill129 replied:
Apparently it didn't, confused as to why since it's been like this through testing. Changed so it lives in a list so it matches the expected form of T0X_arr after self._t is declared

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        else:
+            piece_index = (
+                []
+            )  # iniialise array used to count the number of pieces. Operates by seaching for "A1X_i/T0X_i" and appending i to the array. Can operate if pieces are given out of order.
```
Try to avoid comments on the same line as your code, it makes for nasty formatting.

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+                index
+            ) in (
+                piece_index
+            ):  # looping through each index in order they are given (0 -> n)
```
I'm not sure this is really the order you're getting. Doesn't `np.unique` give you the strings in sorted (lexicographical) order, regardless of the order that they were given?

poneill129 replied:
So it's populating the list `piece_index` with all the unique piecewise parameter keys (i.e. '0000', '0001' etc). Then looks up the corresponding values from valDict using that key. It's a matching process rather than using the iteration number of the loop.

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+            if len(self.piecewise_parameter_information) > 0:
+                # check = hasattr(self,"t")
+                # print(f"Piecewise parameter loader can see t: {check}")
+                if hasattr(self, "_t") is True:
```
Don't use `is True`; just say `if hasattr(...):`

aarchiba replied:
Also if you set this to None in `__init__` you can avoid the need tor `hasattr`.

poneill129 replied:
removed instances of is True. There may still be some `hasattr`'s in other conditions that could follow the same treatment

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+
+    def a1(self):
+        self.A1_val = self.A1X_arr * ls
+        if hasattr(self, "A1X_per_toa") is True:
```
What decides whether `self` has `A1X_per_toa`? When would it happen or not? Can you make sure it's always present?

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        """tt0 = barycentricTOA - T0"""
+        if barycentricTOA is None or self.T0 is None:
+            tt0 = None
+            return tt0
```
Just write `return None`

poneill129 replied:
Written `return None`!

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if barycentricTOA is None or self.T0 is None:
+            tt0 = None
+            return tt0
+        if len(barycentricTOA) > 1:
```
What happens if `len(barycentricTOA)==1`?

poneill129 replied:
Changed the condition such that `len(barycentricTOA)==1` will set correct `T0X_per_toa` array based on if the TOA lies in a group.

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+                )  # Uses the index for each toa array to create arrays where elements are the A1X/T0X to use with that toa
+        if len(barycentricTOA) > 1:
+            check = hasattr(self, "T0X_per_toa")
+            if hasattr(self, "T0X_per_toa") is True:
```
If you're going to create `check`, just use `if check:`; but again, when might `T0X_per_toa` not be present?

poneill129 replied:
Testing variable, deleted.

poneill129 replied:
T0X_per_toa was not present on model initialisation, this was likely a hangover from testing when tt0 gets calculated

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if len(barycentricTOA) > 1:
+            check = hasattr(self, "T0X_per_toa")
+            if hasattr(self, "T0X_per_toa") is True:
+                if len(self.T0X_per_toa) == 1:
```
This `if` statement does the same thing on both branches.

poneill129 replied:
Duplicate branch removed

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+    def d_delayL1_d_par(self, par):
+        if par not in self.binary_params:
+            errorMesg = par + " is not in binary parameter list."
+            raise ValueError(errorMesg)
```
Just construct the error message in the ValueError construction: `raise ValueError(f"{par} is not in binary parameter list.")`

poneill129 replied:
Addressed during string formatting corrections

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if par not in self.binary_params:
+            errorMesg = par + " is not in binary parameter list."
+            raise ValueError(errorMesg)
+        par_obj = getattr(self, par)
```
Is this used anywhere?

poneill129 replied:
Can't tell the exact line this is referencing, but par_obj is used flexibly get the unit we expect with an arbitrary piecewise parameter.

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+        if par not in self.binary_params:
+            errorMesg = par + " is not in binary parameter list."
+            raise ValueError(errorMesg)
+        # print(par)
```
Please deal with debugging `print` statements - either delete them, or if you think they might be useful for debugging, use `log.debug`.

poneill129 replied:
All gone now!

aarchiba commented on src/pint/models/stand_alone_psr_binaries/BT_piecewise.py:
```diff
@@ -0,0 +1,441 @@
+                else:
+                    return np.zeros(len(self.t)) * u.second / par_obj.unit
+
+    def in_piece(self, par):
```
I think this function needs a docstring - what does it do?

poneill129 replied:
Added doc string to in_piece.
It finds which TOAs reference a given piecewise binary parameter group par.
It looks up these values internally from the property group_array_index, which is generated when `self._t` is not `None`
Returns a boolean list (length = number of TOAs), True for the index where TOAs reference a given piecewise parameter, False otherwise. 

Just occurred to me that this may be an issue if the TOAs are reordered between calculations. Unless the TOAs always reach here in a controlled order

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+
+
+@pytest.fixture
+def build_model():
```
Because you use fixtures like `def test_thing(fixture):` ... `y = fixture.attribute`, it makes sense to use a noun-like name; here `model_no_pieces` would make sense, so you'd have `def test_something(model_no_pieces):`.

poneill129 replied:
Addressed this only once (other functions like: build_with_two_pieces could receive a similar treatment)

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    PEPOCH             55352.6375960000000001
+    PLANET_SHAPIRO                          N
+    DM                  14.720549454038943568 1 
+    DM1                                   0.0
```
Can you omit some of the less important bits, like DM1 and PMRA? Keep this a little shorter and easier to glance over? Which bits are actually needed?

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    return toas
+
+
+def get_toa_group_indexes(model, toas):
```
This isn't recognized as a test because it doesn't start with `test_`. Wouldn't it be better to just use `model.which_group_is_toa_in(toas)` directly?

poneill129 replied:
Removed redundant functions like this and similar

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+
+def get_number_of_groups(model):
+    # returns number of groups
+    model.setup()
```
Likewise it's better to use `model.get_number_of_groups()` directly.

poneill129 replied:
Fixed as part of the previous comment

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+@pytest.fixture()
+def build_piecewise_model_with_two_pieces(build_model):
+    # takes the basic model frame and adds 2 non-ovelerlapping pieces to it
+    piecewise_model = deepcopy(build_model)
```
Fixtures always give you a fresh-created new object, so there's no need to copy it.

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    return piecewise_model
+
+
+def test_add_piecewise_parameter(build_model):
```
I think this test is actually checking whether the model round-trips to a par file correctly? If so, maybe call the test something that says that?

aarchiba replied:
There are round-trip tests elsewhere as well, it would be good to add a BTpiecewise model to those, to benefit from that testing code as well.

poneill129 replied:
Unfortunately I couldn't find where some of these additional tests live when updating. Still needed.

Otherwise, renamed test to something appropriate. (Likely other instances of inappropriate test names)

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    copy_param_dict = m3.get_params_dict(which="all")
+    number_of_keys = 0
+    comparison = 0
+    for key, value in param_dict.items():
```
Can you just ask `assert param_dict == copy_param_dict`? If this fails you'll get a list of the mismatched keys/values.

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    comparison = 0
+    for key, value in param_dict.items():
+        number_of_keys = number_of_keys + 1  # iterates up to total number of keys
+        for copy_key, copy_value in param_dict.items():
```
Shouldn't this be `copy_param_dict`?

poneill129 replied:
Good spot, yes it should be

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+
+
+def test_get_number_of_groups(build_piecewise_model_with_two_pieces):
+    # test to make sure number of groups matches with number of added piecewise parameters
```
Can you put this information in the function name?

poneill129 replied:
Unchanged yet, fell through the cracks while updating

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+        np.unique(index, return_counts=True)[1][1],
+    ]
+    expected_toas_in_each_group = [10, 10]
+    is_there_ten_toas_per_group = np.array_equal(
```
it's better to use `numpy.testing.assert_array_equal`, it will tell you more about the mismatch

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    m_piecewise = build_piecewise_model_with_two_pieces
+    toa = make_toas_to_go_with_two_piece_model
+    m_piecewise.setup()
+    rs = pint.residuals.Residuals(toa, m_piecewise)
```
Is this ever used?

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+def test_problematic_group_indexes_and_ranges(build_model):
+    # Test to flag issues with problematic group indexes
+    m_piecewise = build_model
+    with pytest.raises(Exception):
```
Please specify what kind of exception is to be raised.

poneill129 replied:
Included the types of exceptions that should be raised (basically all value errors)

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    # Test to flag issues with problematic group indexes
+    m_piecewise = build_model
+    with pytest.raises(Exception):
+        assert m_piecewise.add_group_range(
```
No point asserting anything if it's supposed to raise an exception. Just run the code.

poneill129 replied:
As suggested, just running the code. New function added to `binary_piecewise.py` called `lock_groups` which runs validate then setup. Avoids the need to run it after every piecewise parameter is added but must be run after adding the a set of/sets of piecewise parameters and date ranges for the boundaries.

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+        assert m_piecewise.add_group_range(
+            m_piecewise.START.value, m_piecewise.FINISH.value, j=-1
+        )
+        assert m_piecewise.add_group_range(
```
This needs to be in a separate `with` statement - the first exception exits the `with`.

poneill129 replied:
Fixed to only include a single assertion in a `with` statement. I realise most of these have been removed during the update

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+        )
+
+
+def add_groups_and_make_toas(build_model, build_piecewise_model_with_two_pieces, param):
```
This should be three separate functions, they don't share any appreciable code.

poneill129 replied:
Separated out into two functions since overlapping groups are now fail validation tests.

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    )
+
+
+def convert_int_into_index(i):
```
Isn't there a utility function to do this in the `BTpiecewise` code?

poneill129 replied:
It's removed now, all it was doing is making 1->0001 so I just declare them where necessary now rather than using a function

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    return truth_array_comparison
+
+
+@pytest.mark.parametrize(
```
You can pass a list of functions to `parametrize`.

poneill129 replied:
Didn't know that, thanks!

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+        "non-complete group coverage",
+    ],
+)
+def test_does_toa_lie_in_group(
```
The three branches here don't share any appreciable code so they should be three separate tests.

poneill129 replied:
Again separated into two tests, since overlapping groups are handled differently now

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    rs_temp = pint.residuals.Residuals(
+        toas, m_piecewise_temp, subtract_mean=False
+    ).resids_value
+    have_residuals_changed = np.invert(rs_temp == rs_value)
```
you can use `rs_temp != rs_value`

poneill129 replied:
Talk about going the long way to get the same thing! Silly oversight on my part

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+
+
+def get_d_delay_d_xxxx(toas, model, parameter_string):
+    rs_temp = pint.residuals.Residuals(toas, model, subtract_mean=False)
```
Do you really need to construct Residuals here? Why?

poneill129 replied:
It appears given the updates I don't (/perhaps never needed to), now constructing residuals is reserved for tests that actually use them.

aarchiba commented on tests/test_BT_piecewise.py:
```diff
@@ -0,0 +1,560 @@
+    return d_delay_temp
+
+
+# @pytest.mark.parametrize("param, index, offset_size",[("T0X",0, 1e-5*u.d),#("T0X",-1, 1e-5*u.d),("A1X",0, 1e-5*ls),("A1X",-1, 1e-5*ls)])
```
You can use `pytest.mark.skip` with a reason to indicate that a test should be skipped.

poneill129 replied:
Thanks for the info, but this test has been removed and replaced now, informed by some of the test suggestions you make in the next comment!

aarchiba commented:
This looks excellent! A little code cleanup would improve it a lot. I'd also like to suggest a few additional tests:

- Can you compare the residuals from BTPiecewise with several pieces to BT with the values appropriate for that piece? This is supposed to give exactly the same residuals but could go wrong. Checking derivatives would be good too.
- There are some tests elsewhere that use `numdifftools` to check analytical derivatives against numerical. Can you integrate a BTPiecewise model into that?
- Can you write some corner-case tests - what happens if the T0X intervals are out of order? If the end is before the beginning? If it's zero-width? If they're duplicated or overlap? Most of these should just raise exceptions, but tests can document what should happen if the user puts something wrong in.
- Can you write a test that creates two different BTpiecewise models and calls various functions of one interlaced with calls to the other? Just to make sure there's no global caching of anything (I'm sure you're fine but it's good to check).
- Can you write some tests that do some calculations, change some values, and do some new calculations; versus doing the second lot of calculations directly? This is to make sure that stuff doesn't linger in BTpiecewise objects from one call to the next.

dlakaplan commented:
Some discussion of the `T2` model came up, and I notice that `T2` implements binary parameter jumps as:
<div class="table-overflow" style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin: 0px 0px 0.5em; padding: 0px; border: 0px; font-family: Merriweather, serif; font-size: inherit; font-stretch: inherit; line-height: inherit; vertical-align: baseline; overflow-x: auto; caret-color: rgb(42, 42, 42); color: rgb(42, 42, 42);">

BPJEP_k | Epoch of a step-jump in the binary parameters |  
-- | -- | --
BPJPH_k | Size of phase jump |  
BTJPB_k | Size of jump in orbital period |  
BTJA1_k | Size of jump in projected semimajor axis |  
BTJECC_k | Size of jump in orbital eccentricity |  
BTJOM_k | Size of jump in longitude of periastron |  
</div>
Is that in any way compatible with how things were implemented here?

scottransom commented:
Hi All:  What is the status of this code?  Is it working?  Or might it be soon?  I have a new student who will be working on timing Ter5A and would love to use this, if possible.

poneill129 commented:
Adding some of my extra thoughts based on Anne's last block of suggestions.

> Can you compare the residuals from BTPiecewise with several pieces to BT with the values appropriate for that piece? This is supposed to give exactly the same residuals but could go wrong. Checking derivatives would be good too.
- Yes! Implemented as: `test_residuals_in_pieces_are_same_as_BT_piecewise` followed by the orbital parameter it is testing.
- Also `test_derivatives_in_pieces_are_same_as_BT_piecewise` followed by the orbital parameter is implemented to check the derivatives correctly being calculated inside/outside of a given piece. Warning: `test_derivatives_in_pieces_are_same_as_BT_piecewise_A1` is producing assertion errors if `assert_array_equal` is used, looks like its some form of round-off issue as the BT and BT_piecewise derivatives are effectively equal.  

>There are some tests elsewhere that use numdifftools to check analytical derivatives against numerical. Can you integrate a BTPiecewise model into that?
Haven't been able to spot where these live. Seems like a good idea to implement still

>Can you write some corner-case tests - what happens if the T0X intervals are out of order? If the end is before the beginning? If it's zero-width? If they're duplicated or overlap? Most of these should just raise exceptions, but tests can document what should happen if the user puts something wrong in.
Implemented as validation tests which are checked in the testing script.

>Can you write a test that creates two different BTpiecewise models and calls various functions of one interlaced with calls to the other? Just to make sure there's no global caching of anything (I'm sure you're fine but it's good to check).
>Can you write some tests that do some calculations, change some values, and do some new calculations; versus doing the second lot of calculations directly? This is to make sure that stuff doesn't linger in BTpiecewise objects from one call to the next.
I think I've made a start on testing some of these in the final test of the script. It performs some basic arithmetic and attempts to change piecewise parameter values in two independent models based on the other model. Currently passing but I doubt this is the exact form you were suggesting. Further discussion required.

poneill129 commented:
Replying to @dlakaplan's comment. Yes there does seem to be a natural way these two can interface. I would need to think about how I would implement this. Either as a `adjust_piecewise_param(new_par_val,par_name)`. I'd need to check how exactly the T2 model adjusts orbital parameters but gut feeling tells me it can be made to work together.

dlakaplan commented:
@poneill129 : are you able to attend the weekly PINT call to discuss this?

poneill129 commented:
> @poneill129 : are you able to attend the weekly PINT call to discuss this?

Yes I will be, will this be on Thursday? If so, what time does it start?

abhisrkckl commented:
A completed version of this (#1545) has already been merged. So I am closing this.
